server/channels/app/mention_keywords.go

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package app
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/mattermost/mattermost/server/public/model"
+)
+
+const (
+	mentionableUserPrefix  = "user:"
+	mentionableGroupPrefix = "group:"
+)
+
+type MentionableID string
+
+func MentionableUserID(userID string) MentionableID {
+	return MentionableID(fmt.Sprint(mentionableUserPrefix, userID))
+}
+
+func MentionableGroupID(groupID string) MentionableID {
+	return MentionableID(fmt.Sprint(mentionableGroupPrefix, groupID))
+}
+
+func (id MentionableID) AsUserID() (userID string, ok bool) {
+	idString := string(id)
+	if !strings.HasPrefix(idString, mentionableUserPrefix) {
+		return "", false
+	}
+
+	return idString[len(mentionableUserPrefix):], true
+}
+
+func (id MentionableID) AsGroupID() (groupID string, ok bool) {
+	idString := string(id)
+	if !strings.HasPrefix(idString, mentionableGroupPrefix) {
+		return "", false
+	}
+
+	return idString[len(mentionableGroupPrefix):], true
+}
+
+type MentionKeywords map[string][]MentionableID
+
+func (k MentionKeywords) AddUser(profile *model.User, channelNotifyProps map[string]string, status *model.Status, allowChannelMentions bool) MentionKeywords {
+	mentionableID := MentionableUserID(profile.Id)
+
+	userMention := "@" + strings.ToLower(profile.Username)
+	k[userMention] = append(k[userMention], mentionableID)
+
+	// Add all the user's mention keys
+	for _, mentionKey := range profile.GetMentionKeys() {
+		// note that these are made lower case so that we can do a case insensitive check for them
+		mentionKey = strings.ToLower(mentionKey)
+
+		if mentionKey != "" {
+			k[mentionKey] = append(k[mentionKey], mentionableID)
+		}
+	}
+
+	// If turned on, add the user's case sensitive first name
+	if profile.NotifyProps[model.FirstNameNotifyProp] == "true" && profile.FirstName != "" {
+		k[profile.FirstName] = append(k[profile.FirstName], mentionableID)
+	}
+
+	// Add @channel and @all to k if user has them turned on and the server allows them
+	if allowChannelMentions {
+		// Ignore channel mentions if channel is muted and channel mention setting is default
+		ignoreChannelMentions := channelNotifyProps[model.IgnoreChannelMentionsNotifyProp] == model.IgnoreChannelMentionsOn || (channelNotifyProps[model.MarkUnreadNotifyProp] == model.UserNotifyMention && channelNotifyProps[model.IgnoreChannelMentionsNotifyProp] == model.IgnoreChannelMentionsDefault)
+
+		if profile.NotifyProps[model.ChannelMentionsNotifyProp] == "true" && !ignoreChannelMentions {
+			k["@channel"] = append(k["@channel"], mentionableID)
+			k["@all"] = append(k["@all"], mentionableID)
+
+			if status != nil && status.Status == model.StatusOnline {
+				k["@here"] = append(k["@here"], mentionableID)
+			}
+		}
+	}
+
+	return k
+}
+
+func (k MentionKeywords) AddUserID(userID string, keyword string) MentionKeywords {
```

Comment:
The only time in this PR where we call this, it seems we have the `user` in hand and `keyword` is just the `username`:
```go
channelKeywords[c.Id].AddUserID(userID, "@"+user.Username)
```

Can we just make this method `AddUser(user)` like `AddGroup(group)` below?

Reply:
I could make it `AddUsername` instead, but the name `AddUser` is already taken by a method that adds all of the user's mention keys.

Come to think of it, I wonder if there's a bug in the code that uses this where the persistent notifications aren't case insensitive because this code doesn't `strings.ToLower` the username currently. That's outside the scope for this though

Reply:
`AddUsername` would bring a lot of clarity for me -- I kept stumbling over /what/ was being added.